### DISCUSSION TITLE
all reports rendered via `tar_quarto_rep` show the same figures [help]

### DISCUSSION THREAD

**Discussion by tklebel** [ropensci/tarchetypes#157] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I am rendering parameterized reports with `tar_quarto_rep`. The basic steps of parameterising all work fine (I am able to use the parameters that I supply via `tar_quarto_rep`). However, when I render the reports which include figures, all reports show the same figures; that is, the figures from the last run (in this case with `disp` on the x-axis). In the code below, an output directory `report_files` is created, and both reports link to the figures of this directory. I thus might have `n` reports, but only `one` set of figures. I am not sure whether I am misunderstanding something, or whether this is a bug.

``` r
targets::tar_dir({ # tar_dir() runs code from a temporary directory.
  # Parameterized Quarto:
  lines <- c(
    "---",
    "title: 'report.qmd file'",
    "output_format: html_document",
    "params:",
    "  par: variable:",
    "    \"mpg\"",
    "---",
    "This report is rendered with variable = `r params$variable`.",
    "```{r}",
    "library(ggplot2)",
    "ggplot(mtcars, aes(.data[[params$variable]], hp)) +",
    "  geom_point()",
    "```"
  )
  writeLines(lines, "report.qmd")

  targets::tar_script({
    library(tarchetypes)
    list(
      tar_quarto_rep(
        report,
        path = "report.qmd",
        execute_params = tibble::tibble(variable = c("mpg", "disp"))
      )
    )
  }, ask = FALSE)
})
```

<sup>Created on 2023-11-23 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

For completeness, these are the two files I use (might be easier to copy-paste):
`_targets.R`:
```r
library(tarchetypes)

params <- tibble::tibble(
  variable = c("mpg", "disp")
)

list(tar_quarto_rep(report, path = "report.qmd", execute_params = params))
```

`report.qmd`: 

```r
---
title: "test"
format: html
params:
  variable:
    "mpg"
---

This report is rendered with variable = `r params$variable`.

```{r}
library(ggplot2)

ggplot(mtcars, aes(.data[[params$variable]], hp)) +
  geom_point()
# insert three backticks
```

**Comment by tklebel** [comment]
I was able to solve the issue by creating self-contained html files:

```yaml
---
title: "test"
format: 
  html: 
    self-contained: true
params:
  variable:
    "mpg"
---
```

Maybe this is something to add to the documentation, at least?

**Reply by tklebel** [reply]
Unfortunately this did not entirely solve the problem, because my reports are in subdirectories. Rendering reports from subdirectories works in principle (see https://github.com/ropensci/tarchetypes/issues/144), but fails when one sets `self-contained: true`. The report still renders, but is missing CSS and links to figures. This is a separate issue though (happy to file a proper issue once the maintainer is back from vacation).

**Reply by wlandau** [reply]
You might need to parameterize the R Markdown figure path with something like `knitr::opts_chunk$set(fig.path = paste0("figures-", params$variable))`.

**Reply by tklebel** [reply]
Thanks for pointing this out, I had not thought of it! After a slight modification they end up in the correct directory:

```r
knitr::opts_chunk$set(fig.path = paste0("report_files/figures-", params$variable, "/"))
``` 

Is this already mentioned somewhere in the documentation? I think it would be very helpful to add it, at least to `tar_quarto_rep`.

**Reply by wlandau** [reply]
> After a slight modification they end up in the correct directory:

Glad that works for you.

> Is this already mentioned somewhere in the documentation? I think it would be very helpful to add it, at least to tar_quarto_rep.

There are many edge cases and gotchas like this in literate programming, too many for the manual or help files.
